PEP 767: Annotating Read-Only Attributes (#4127)
python/peps: peps/pep-9999.rst

> PEP: 9999

Reviewer: You can use PEP 767.

> Title: Classes & protocols: Read-only attributes

Reviewer: ```suggestion
Title: Read-only attributes for classes and protocols
```

Author: How about
```suggestion
Title: Read-only class attributes
```
?

Author: "Annotating Read-Only Attributes"
(also just noticed many PEPs follow Title-Case)

> Author: <REQUIRED: list of authors' real names and optionally, email addrs>

Reviewer: I guess we forgot to update the "real names" part here. It's fine to say something like "Enegg <your@email.com>" (#4058). I'll look into updating the template.

> Sponsor: Carl Meyer <carl@oddbird.net>

Reviewer 2: @carljm Please can you confirm you are sponsoring?

Discussions-To: <REQUIRED: URL of current canonical discussion thread>
Status: Draft
Type: Standards Track
Topic: Typing

> Created: ?-Nov-2024

Reviewer: ```suggestion
Created: 18-Nov-2024
```

> Python-Version: 3.?

Reviewer: ```suggestion
Python-Version: 3.14
```

Abstract
========

:pep:`705` introduced the :external+py3.13:data:`typing.ReadOnly` type qualifier
to allow defining read-only :class:`typing.TypedDict` items.

This PEP proposes expanding the scope of ``ReadOnly`` to class and protocol
attributes, as a single concise way to mark them read-only. Some parity changes
are also made to :data:`typing.Final`.

Akin to :pep:`705`, it makes no Python grammar changes. Correct usage of
read-only attributes is intended to be enforced only by static type checkers.

Motivation
==========

Python type system lacks a single concise way to mark an :term:`attribute` read-only.

> This feature is common in other object-oriented languages (such as `C# <https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/readonly>`_),

Reviewer: To establish that it is "common", would be useful to mention one or two other languages. In particular, TypeScript (which has `readonly`) would be nice to mention as it also has a gradual type system.

Author: Perhaps wording it as
```suggestion
This feature is present in other object-oriented languages (such as `C# <https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/readonly>`_),
```
would be better; claiming it is common is somewhat bold

Reviewer 3: I think that's fine, but either way mentioning TypeScript would be useful!

and is useful for restricting attribute mutation at a type checker level, as well
as defining a broad interface for structural subtyping.

.. _classes:

Classes
-------

> There are 3 major ways of achieving read-only attributes, honored by type checkers:

Reviewer: ```suggestion
There are three major ways of achieving read-only attributes, honored by type checkers:
```

* annotating the attribute with :data:`typing.Final`::

    class Foo:
        number: Final[int]

        def __init__(self, number: int) -> None:
            self.number = number

    class Bar:
        def __init__(self, number: int) -> None:
            self.number: Final = number

  - Supported by :mod:`dataclasses` (and type checkers since `typing#1669 <https://github.com/python/typing/pull/1669>`_).
  - Overriding ``number`` is not possible - the specification of ``Final``

>     imposes the name cannot be overridden in subclasses.

Reviewer: ```suggestion
    imposes that the name cannot be overridden in subclasses.
```

            self._number = number

        @property
        def number(self) -> int:
            return self._number

  - Overriding ``number`` is possible. *Type checkers disagree about the specific rules*. [#overriding_property]_
  - Read-only at runtime. [#runtime]_
  - Requires extra boilerplate.
  - Supported by :mod:`dataclasses`, but does not compose well - the synthesized
    ``__init__`` and ``__repr__`` will use ``_number`` as the parameter/attribute name.

* using a "freezing" mechanism, such as :func:`dataclasses.dataclass` or :class:`typing.NamedTuple`::

    @dataclass(frozen=True)
    class Foo:
        number: int

    class Bar(NamedTuple):
        number: int

  - Overriding ``number`` is possible in the ``@dataclass`` case.
  - Read-only at runtime. [#runtime]_
  - No per-attribute control - these mechanisms apply to the whole class.
  - Frozen dataclasses incur some runtime overhead.
  - ``NamedTuple`` is still a ``tuple``. Most classes do not need to inherit
    indexing, iteration, or concatenation.

.. _protocols:

Protocols
---------

Paraphrasing `this post <https://github.com/python/typing/discussions/1525>`_,

> there is no way of defining a :class:`~typing.Protocol`, such that the only

Reviewer 3: ```suggestion
there is no way of defining an attribute `x: T` on a :class:`~typing.Protocol`, such that the only
```

> there is no way of defining a :class:`~typing.Protocol`, such that the only

Author: ```suggestion
there is no way of defining an attribute ``name: T`` on a :class:`~typing.Protocol`, such that the only
```

requirements to satisfy are:

> 1. ``hasattr(obj, name)``

Reviewer: ```suggestion
1. ``hasattr(obj, "name")``
```
Or use `getattr(obj, name)` on the next line. Right now the two are inconsistent.

2. ``isinstance(obj.name, T)`` [#invalid_typevar]_

The above are satisfiable at runtime by all of the following:

1. an object with an attribute ``name: T``,
2. a class with a class variable ``name: ClassVar[T]``, [#invalid_typevar]_
3. an instance of the class above,
4. an object with a ``@property`` ``def name(self) -> T``,
5. an object with a custom descriptor, such as :func:`functools.cached_property`.

Note that the attribute being marked ``Final`` or the property defining a setter
do not impact this.

The most common practice is to define such a protocol with a ``@property``::

    class HasName(Protocol):
        @property
        def name(self) -> T: ...

Type checkers special-case this definition, such that objects with plain attributes
are assignable to the type. However, instances with class variables and descriptors

> other than ``property`` are rejected.

Reviewer: Is this consistently implemented? It doesn't appear to be specified in detail, though [the spec](https://typing.readthedocs.io/en/latest/spec/protocol.html) says "To define a read-only protocol variable, one can use an (abstract) property."

Author: Huh. I have been relying on what pyright told me the whole time; turns out this is not true.
[mypy](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=12d556bb6ef4a9a49ff4ed4776604750), [pyre](https://pyre-check.org/play?input=%23%20pyre-strict%0Afrom%20abc%20import%20abstractmethod%0Afrom%20functools%20import%20cached_property%0Afrom%20typing%20import%20ClassVar%2C%20Protocol%2C%20final%0A%0A%0Aclass%20HasFoo(Protocol)%3A%0A%20%20%20%20%40property%0A%20%20%20%20%40abstractmethod%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20...%0A%0A%0A%23%20assignability%0A%0A%0Aclass%20FooAttribute%3A%0A%20%20%20%20foo%3A%20int%0A%0Aclass%20FooProperty%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooClassVar%3A%0A%20%20%20%20foo%3A%20ClassVar%5Bint%5D%20%3D%200%0A%0Aclass%20FooDescriptor%3A%0A%20%20%20%20%40cached_property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooPropertyCovariant%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20bool%3A%20return%20False%0A%0Aclass%20FooInvalid%3A%0A%20%20%20%20foo%3A%20str%0A%0Aclass%20NoFoo%3A%0A%20%20%20%20bar%3A%20str%0A%0A%0Aobj%3A%20HasFoo%0Aobj%20%3D%20FooAttribute()%20%20%23%20ok%0Aobj%20%3D%20FooProperty()%20%20%20%23%20ok%0Aobj%20%3D%20FooClassVar%20%20%20%20%20%23%20ok%0Aobj%20%3D%20FooClassVar()%20%20%20%23%20ok%0Aobj%20%3D%20FooDescriptor()%20%23%20ok%0Aobj%20%3D%20FooPropertyCovariant()%20%23%20ok%0Aobj%20%3D%20FooInvalid()%20%20%20%20%23%20err%0Aobj%20%3D%20NoFoo()%20%20%20%20%20%20%20%20%20%23%20err%0Aobj%20%3D%20None%20%20%20%20%20%20%20%20%20%20%20%20%23%20err%0A%0A%0A%23%20explicit%20impl%0A%0A%0Aclass%20FooAttributeImpl(HasFoo)%3A%0A%20%20%20%20foo%3A%20int%0A%0Aclass%20FooPropertyImpl(HasFoo)%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooClassVarImpl(HasFoo)%3A%0A%20%20%20%20foo%3A%20ClassVar%5Bint%5D%20%3D%200%0A%0Aclass%20FooDescriptorImpl(HasFoo)%3A%0A%20%20%20%20%40cached_property%0A%20%20%20%20def%20foo(self)%20-%3E%20int%3A%20return%200%0A%0Aclass%20FooPropertyCovariantImpl(HasFoo)%3A%0A%20%20%20%20%40property%0A%20%20%20%20def%20foo(self)%20-%3E%20bool%3A%20return%20False%0A%0Aclass%20FooInvalidImpl(HasFoo)%3A%0A%20%20%20%20foo%3A%20str%0A%0A%40final%0Aclass%20NoFooImpl(HasFoo)%3A%0A%20%20%20%20bar%3A%20str%0A) both allow `ClassVar`s and descriptors, only [pyright](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBAhgRgYygSwgBzCALrOBnLEGBLCAUywAswATAKFEimAFcA7EsMAGzxXUw4ExSmRoB9NODRlsATwbhoWOWmRsA5vwzYoAYW4w8eAGowQAGigAFcFjAIeV4Opjc6HhIeNQAEkYAxLgAKWzB7R24ASgAuOigEqAABKTAZeXjEpPgCIhJyKlpMhJoyYGYQvDJuYCioAFoAPhQ2LBioADoujzoAYlhjZA02eGRuZBUeryM%2BILAAQSxCZDgWLDI4xIqwdvUsT29ZrjD0lU2s1NOFLdLy4Erq2obmvfaQChYQNigABgOZqBzAwzMwgc4Je47fSHUEAbT2AF0oABeX7-HxzAAiZDwCBAyDQ9jBxWSwgQogkl1kkxuZW2wSqNTqTRabSg7ywn2%2Bfzo0wxx2k1LkejAADdzMgYK1wckqRlaXcHkznlA4FxuG8Pl9AW4quijmAAJJscXjGgyyHtXL6qAAOTAcxlcHMVsIPTAcAAVu1-Hg5nQPZ6UYCuItlqt1sE6lB%2BmAANYBr3BuYnIVRxKxhOB5NcYHGUFbGNQeOJoOooEw8zphKZ0s5sDY3H4wmYdO17PlgVpIUi8X4qVYNvFrNJztGk1uZA0as1qCyEB11H2uYzwtF%2BeLu1gNhkNdr-objwHgAeaHGCAm2ncNrmYfxEbIhvQ3GCvrmsRJltZN67VyfZ9fQIuA-LYUkFeVEluelGSeFlXnZLVuR-MA81Mcx-xfN9gItLh2lQuFEWDHk%2BQNRs8QJIkMMAv1sJJJIyQpSRwJpSC6UhBlHmZF5pQQzltWIw4QzAVN5F7CUByorCwBAi5mOuVjFTADjlRZNUeE1PjvgCXUyGQ41TSnSSgOknCoWtOgkhcEZ3BIrc5iMmiTJJZ0wSga0gA) rejects them.
They also support explicitly declaring the implementation via inheritance.

This changes the whole narrative, which is troubling. It would seem that only a minor update to the spec is required so as to have all type checkers consistent here

Covering the extra possibilities induces a great amount of boilerplate, involving
creation of an abstract descriptor protocol, possibly also accounting for
class and instance level overloads.
Worse yet, all of that is multiplied for each additional read-only attribute.

Rationale
=========

These problems can be resolved by an attribute-level :external+typing:term:`type qualifier`.
``ReadOnly`` has been chosen for this role, as its name conveys the intent well,

> and the newly proposed changes complement its semantics defined in the :pep:`705`.

Reviewer: ```suggestion
and the newly proposed changes complement its semantics defined in :pep:`705`.
```

> A class with a read-only instance attribute can be now defined as such::

Reviewer: ```suggestion
A class with a read-only instance attribute can now be defined as::
```

    from typing import ReadOnly

    class Member:
        def __init__(self, id: int) -> None:
            self.id: ReadOnly = id

...and a protocol described in :ref:`protocols` is now just::

    from typing import Protocol, ReadOnly

    class HasName(Protocol):
        name: ReadOnly[str]

    def greet(obj: HasName, /) -> str:
        return f"Hello, {obj.name}!"

* A subclass of ``Member`` can redefine ``.id`` as a writable attribute or a
  :term:`descriptor`. It can also :external+typing:term:`narrow` the type.
* The ``HasName`` protocol can be implemented by any mechanism allowing for ``.name`` access.
* The ``greet`` function can now accept a wide variety of compatible objects,
  while being explicit about no modifications being done to the input.

Specification
=============

The :external+py3.13:data:`typing.ReadOnly` :external+typing:term:`type qualifier`
becomes a valid annotation for :term:`attributes <attribute>` of classes and protocols.

> It is used to indicate that an attribute should not be reassigned or ``del``\ eted.

Author: (fixing backticks to add this in a batch)
```suggestion
Type checkers should error on any attempt to reassign or ``del``\ ete an attribute annotated with ``ReadOnly``.
```

The deletability rule should be extended to ``Final`` as well, as it is currently
not specified.

Akin to ``Final``, read-only attributes do not influence the mutability of
the assigned object. Immutable ABCs and containers may be used in combination with

> ``ReadOnly`` to prevent mutation of such values.

Author: This comes from https://typing.readthedocs.io/en/latest/spec/qualifiers.html#semantics-and-examples
> Note that declaring a name as final only guarantees that the name will not be re-bound to another value, but does not make the value immutable. Immutable ABCs and containers may be used in combination with Final to prevent mutating such values: [...]

Unfortunately it's fairly deep into that section :/

> For example, given a `ReadOnly` member `names` of an object `x`, `x.names.append("Pat")`
is not an error.

There's an example of a mutable read-only attribute later in the `Initialization` section, I suppose I could float it closer to this paragraph

Syntax
------

``ReadOnly`` can be used at class-level or within ``__init__`` to mark individual
attributes read-only:

.. code-block:: python

    class Book:
        id: ReadOnly[int]

        def __init__(self, id: int, name: str) -> None:
            self.id = id
            self.name: ReadOnly = name

The explicit type in ``ReadOnly[<type>]`` can be omitted if the declaration has
an initializing value. A type checker should apply its usual type inference
rules to determine the type of ``name``.

> If an attribute is already implied to be read-only, like in the frozen :ref:`classes`,

Reviewer: ```suggestion
If an attribute is already implied to be read-only, like in frozen :ref:`classes`,
```

explicit declarations should be permitted and seen as equivalent, except ``Final``
additionally forbids overriding in subclasses:

.. code-block:: python

    @dataclass(frozen=True)
    class Point:
        x: ReadOnly[int]

>         y: Final[int]

Reviewer: Could use an example where a child class overrides the ReadOnly field with a narrower type. Also, might be worth commenting on the fact that ReadOnly is redundant here, as all fields are implicitly read-only.

times the attribute can be assigned to.

.. code-block:: python

    from collections import abc
    from typing import ReadOnly

    class Band:
        name: str
        songs: ReadOnly[list[str]]

        def __init__(self, name: str, songs: abc.Iterable[str] | None = None) -> None:
            self.name = name
            self.songs = []

            if songs is not None:
                # multiple assignments during initialization are fine
                self.songs = list(songs)

        def clear(self) -> None:
            # error: assignment to read-only "songs" outside initialization
            self.songs = []

    band = Band(name="Bôa", songs=["Duvet"])
    band.name = "Python"  # ok: "name" is not read-only
    band.songs = []  # error: "songs" is read-only
    band.songs.append("Twilight")  # ok: list is mutable

    class SubBand(Band):
        def __init__(self) -> None:
            # error: cannot assign to a read-only attribute of base class
            self.songs = []

An initializing value at a class level can serve as a `flyweight <https://en.wikipedia.org/wiki/Flyweight_pattern>`_
default for instances:

.. code-block:: python

    class Patient:

>         number: ReadOnly = 0

Reviewer 4: I think we should require `ReadOnly` to be explicitly parameterized in this case. Otherwise, a type checker has to look at `__init__` to figure out the type.

Reviewer 4: I was worried about cases where the immediate initialization might not give complete information about the type (e.g.,
```
class C:
    x: ReadOnly = 0
    def __init__(self):
        self.x = 4.2
```
but if it's specified that the type is always inferred only from the RHS of the assignment regardless, I suppose that works.

Reviewer 5: So are we trying to say that A is allowed and B is not allowed?

A:
```python
class C:
    x: Readonly[int] = 2

    def __init__(self):
        if random() < 0.1:
            self.x = 3
```

B:
```python
class C:
    x: Readonly = 2

    def __init__(self):
        if random() < 0.1:
            self.x = 3
```

(Because in B, the type could be inferred as `Literal[2]` so `3` would be invalid.)

Whatever we're saying, it could be made more clear.

Reviewer 4: The current draft would allow both A and B. My suggestion was that we not allow B.

Reviewer 3: There are several different ways that the code example in B could be "not allowed". One is that bare `ReadOnly` is not allowed as an annotation anywhere, another is that bare `ReadOnly` is not allowed as a class attribute annotation, and a third is that it is allowed here, but the type checker infers the type `Literal[2]`, and so the assignment of `3` in `__init__` is a type error.

The current draft allows the annotation, and leaves the inference unspecified, so the annotation in (B) would be accepted, and whether the assignment in `__init__` is accepted would depend on how the type checker chooses to do inference.

I (weakly) favor side-stepping all of this by just making bare `ReadOnly` never allowed anywhere. I understand the appeal of not requiring annotations that can be inferred, I just think that once you're writing a `ReadOnly` annotation, it's not that hard to specify the type, too, and it makes the behavior much less type-checker-dependent.

I think it will be a bit more confusing if bare `ReadOnly` is allowed in some places but not in others.

Author: I made it required everywhere.

        def __init__(self, number: int | None = None) -> None:
            if number is not None:
                self.number = number

This feature should also be supported by ``Final`` attributes. Specifically,
``Final`` attributes initialized in a class body **should no longer** imply ``ClassVar``,

> and should remain assignable to within ``__init__``.

Reviewer 6: What is the motivation for this change? I'm opposed to changing the semantics of final, since this seems to have some problematic side effects. Currently class-level final attributes can be used to define constants, where the literal value is known statically:
```py
class Foo:
    FOO: Final = 5
    BAR: Final = 6
    
    def __init__(...): ...

x: Literal[5] = Foo.FOO  # OK
```
(Mypyc even treats `Foo.FOO` as a compile-time constant at runtime, replacing references to `Foo.FOO` with 5.)

If we make the change as suggested in the PEP draft, it seems to imply that one of these is true:
1. To infer the type of a class-level final attribute, a type checker will need to combine types from assignments in two scopes: the class body and the `__init__` method.
2. The code above would start generating a type error (since the inferred type of `Foo.FOO` would likely have to be `int` if we must allow an arbitrary `__init__`), breaking backward compatibility.

I think (2) above is not reasonable due to the backward compatibility break. 

I also argue that (1) would also be quite a drastic change, since it adds "spooky action at a distance" where two assignments in two different parts of a file impact the type of an attribute. Here's what I'm thinking about:
```py
class Foo
    foo: Final = 5  # No longer can use initializer to infer type

    # assume lots of other stuff
    # ...
    
    def __init__(self, foo: int | None) -> None:
        if foo is not None:
            self.foo = foo  # This assignment means that "foo" has type "int"
```
I don't think there is a precedent for the above kind of type inference behavior, and it would be quite complex to implement in mypy at least. Also, to preserve backward compatibility, it would require new kind of type inference logic at least in mypy, again because there is no precedent for this kind of type inference behavior. This seems unnecessarily complicated considering that `ReadOnly` seems to be a better fit for the above use case.

In summary, I think that this change seems out of place in the PEP, and increases implementation complexity significantly without a clear benefit that I can see.

Author: I didn't initially plan on touching `Final`. I might have got the wrong idea, but @erictraut's initial [feedback](https://discuss.python.org/t/expanding-readonly-to-normal-classes-protocols/67359/45) seemed to suggest changing the spec of `Final` to allow assignment at class level and within `__init__` [^1]:
> I’m not convinced that we should allow double initialization — in both the class body and in __init__. Is there a good reason to allow both initialization techniques within the same class? Interestingly, mypy allows this for Final but pyright does not. The typing spec is currently silent on whether this should be allowed. Regardless of where we land on this decision, it would be good for this PEP to provide clarity for both ReadOnly and Final.

...hence the change. Likely should've discussed this more thoroughly.

[^1]: Under the precondition that `ReadOnly` obtains the feature

Reviewer 3: Eric's feedback definitely did suggest clarifying the spec for `Final` in this PEP, though it didn't advocate for any particular conclusion or change.

My feeling is that we don't need to change the behavior of `Final` to make it the same as `ReadOnly`; that causes churn, and I think it's OK if they behave differently. I see this difference as deriving naturally from their different meaning; `Final` is really supposed to mean "this is the only assignment to this name, it is final", whereas `ReadOnly` just means "once an instance of this class is constructed, this attribute can't be assigned to."

I (weakly) feel also that it's a bit out of scope for this PEP to clarify existing underspecified aspects of `Final`, when those could just as well be clarified in a separate PR to the typing spec, approved by the typing council. But I'm not strongly opposed to this. If we do it, I would suggest collecting everything in this PEP related to `Final` in a single section, which presents a comparison of the semantics of `Final` and `ReadOnly`, and any clarifications we want to make to the specification of `Final`, all together.

.. note::
    Classes defining :data:`~object.__slots__` cannot make use of this feature.
    An attribute with a class-level value cannot be included in slots,
    effectively making it a class variable.
    Type checkers may warn or suggest explicitly marking the attribute as a ``ClassVar``.

> Type checkers should warn on read-only attributes which may be left uninitialized

Reviewer 5: ```suggestion
Type checkers may warn on read-only attributes which may be left uninitialized
```
Making sure that attributes are initialized is not an easy problem. While Python has not really been trying to do it much, Typescript has been trying to do it for a long time. A search in TypeScript issues for "strictPropertyInitialization" yields 98 open issues and 540 closed issues.

TypeScript has an annotation to assert that something is initialized when the type checker can't see it, and Python doesn't, so Python is less-equipped to handle this.

I would like it if type checkers started trying to move towards this. But I think making it a requirement in this PEP is not the right way to push towards it.

.. code-block:: python

    class Foo:
        foo: ClassVar[ReadOnly[str]] = "foo"
        bar: Annotated[ReadOnly[int], Gt(0)]

.. code-block:: python

    class Foo:
        foo: ReadOnly[ClassVar[str]] = "foo"
        bar: ReadOnly[Annotated[int, Gt(0)]]

This is consistent with the interaction of ``ReadOnly`` and :class:`typing.TypedDict`
defined in :pep:`705`.

``ClassVar`` excludes read-only attributes from being assignable to within
initialization methods.

Rules of ``Final`` should take priority when combined with ``ReadOnly``. As such,
type checkers may warn on the redundancy of combining the two type qualifiers.

Backwards Compatibility
=======================

This PEP introduces new contexts where ``ReadOnly`` is valid. Programs inspecting
those places will have to change to support it. This is expected to mainly affect type checkers.

However, caution is advised while using the backported ``typing_extensions.ReadOnly``
in older versions of Python. Mechanisms inspecting annotations may behave incorrectly
when encountering ``ReadOnly``; in particular, the ``@dataclass`` decorator
which `looks for <https://docs.python.org/3/library/dataclasses.html#class-variables>`_

> ``ClassVar`` will incorrectly treat ``ReadOnly[ClassVar[...]]`` as an instance attribute.

Reviewer: Add a recommendation that users write `ClassVar[ReadOnly[...]]` to avoid this?

Author: Sure.
I'd assume the wording of
> We recommend that in such circumstances `ReadOnly` is placed as the inner type qualifier, like `ClassVar[ReadOnly[...]]`

Is it ok to use "we" (the authors) within the PEP? Or perhaps "It is recommended ..." would be better?

Reviewer 3: "Authors should prefer `ClassVar[ReadOnly[...]]` over `ReadOnly[ClassVar[...]]`..." ?

Reviewer: I'd write something like "To avoid issues with runtime introspection, use `ClassVar[ReadOnly[...]]` instead of `ReadOnly[ClassVar[...]]`.

Author: I assumed the recommendation should apply in the general case 🤔
"To avoid issues with introspection, prefer nesting `ReadOnly` within other type qualifiers - use `ClassVar[ReadOnly[...]]` instead of `ReadOnly[ClassVar[...]]`."
Though it might give the wrong idea that you're always required to nest the type, or that one should put it within a generic that is not a type qualifier

Security Implications
=====================

There are no known security consequences arising from this PEP.

How to Teach This
=================

[How to teach users, new and experienced, how to apply the PEP to their work.]

Open Issues
===========

Assignment in ``__new__``
-------------------------

Immutable classes like :class:`fractions.Fraction` often do not define ``__init__``;
instead, they perform initialization in ``__new__`` or classmethods. The proposed
feature won't be useful to them. 

> OTOH, allowing assignment within ``__new__`` (and/or classmethods) could open way

Reviewer: Expand this abbreviation

Author: I have assumed I could be less formal in the `Open Issues` section, my bad 😄

to non-trivial bugs:

.. code-block:: python

    class Foo:
        # fully initialized objects
        object_cache: ReadOnly[ClassVar[dict[int, Self]]] = {}

        foo: ReadOnly[int]

        def __new__(cls, foo: int) -> Self:
            if foo + 1 in cls.object_cache:
                # this instance is already initialized
                self = cls.object_cache[foo + 1]

            else:
                # this instance is not
                self = super().__new__(cls)

            # assignment to an object which has been initialized before,
            # breaking the invariant a read-only attribute can be assigned to
            # only during its initialization?
            self.foo = foo

            cls.object_cache[foo] = self
            return self

To my understanding, properly detecting this problem would require type checkers

> to keep track of the "level of initialization" of an object.

Reviewer 3: As discussed on Discourse, I think in practice this could be simplified to "was the object directly acquired from a call to super-type `__new__`", which I think is reasonable to track.

And it could also be an option to allow type checker to accept this unsoundness, in exchange for a simpler implementation.

This issue doesn't seem to impact ``__init__``, since it's rather uncommon to
ever rebind ``self`` within it to any other object, and type checkers could

> flag the action as whole.

Reviewer 3: This should probably also mention "calling `__init__` on an existing instance" as another action that would wrongly bypass readonly, if not flagged by type checkers.

Reviewer: @Enegg I think this is still unresolved.

Author: Right. @carljm, since this issue has been removed, do you think there's still value in including what you said?
If yes, it'd probably be a note in the `Initialization` section, or perhaps we should view it as an Open Issue

Reviewer 3: I think it would still be useful to note, in the Initialization section, that if type checkers allow explicit re-calling of the `__init__` method on an instance, this opens the door for a `ReadOnly` attribute to be reassigned on an existing instance.

Personally I would be OK with strengthening this to "type checkers should not allow re-calling `__init__` on an existing instance", but I would want to get feedback from @erictraut on that, since pyright currently does allow it.

Reviewer 7: When you say "re-calling `__init__`", do you mean something like `super().__init__(<args>)`? This is a common pattern, so I don't think it should be disallowed in general. Are you suggesting that it be disallowed only if a `ReadOnly` field is present in one of the class' ancestors? I'm not in favor of mandating such a restriction. I'd want to see reasonably strong evidence that this is a common source of bugs before going down that road.

Reviewer 3: @erictraut Sorry, no, this is what I mean:

```py
class A:
    def __init__(self, val: int) -> None:
        self.val: ReadOnly[int] = val

a = A(3)
a.__init__(4)
```

I think this is very rarely used, and it's already unsound with inheritance because we don't enforce LSP on `__init__`, so I think it shouldn't be allowed. But it's also probably rare enough that it doesn't matter too much that we require disallowing it.

Reviewer 7: Ah, I see what you mean. 

IMO, I don't think this is a case that's important enough to cover in the typing spec one way or another. If someone wants to call `__init__` manually on an object, they're already treading on thin ice and had better know what they're doing.

From my perspective, it's not the job of a type checker to protect developers from all possible ways of misusing or abusing the language. This particular case feels more like a linter rule than something a type checker should check for.

Reviewer 3: I don't feel strongly about specifying it. I don't even feel strongly about mentioning it in this PEP. Happy with whatever you prefer, @Enegg.

Extending initialization
------------------------

Mechanisms such as :func:`dataclasses.__post_init__` or attrs' `initialization hooks <https://www.attrs.org/en/stable/init.html#hooking-yourself-into-initialization>`_
augment initialization by providing a set of dunder hooks which will be called
once during instance creation. The current rules would disallow assignment in those

> hooks. Specifying any single method in the pep isn't enough, as the naming and

Reviewer: ```suggestion
hooks. Specifying any single method in the PEP isn't enough, as the naming and
```

functionality differs between mechanisms (``__post_init__`` vs ``__attrs_post_init__``).

``ReadOnly[ClassVar[...]]`` and ``__init_subclass__``
-----------------------------------------------------

Should this be allowed?

.. code-block:: python

    class URI:
        protocol: ReadOnly[ClassVar[str]] = ""

        def __init_subclass__(cls, protocol: str = "") -> None:
            cls.foo = protocol

    class File(URI, protocol="file"): ...

``Final`` in protocols
----------------------

It's been `suggested <https://discuss.python.org/t/expanding-readonly-to-normal-classes-protocols/67359/45>`_

> to clarify in this PEP whether ``Final`` should be supported by protocols.

Reviewer 3: It would be good to clear this up, but it seems out of scope for this PEP, and like something that could be handled with a PR to the spec and approval by the typing council; it doesn't need a PEP.

Footnotes
=========

.. [#overriding_property]
    Pyright in strict mode disallows non-property overrides.
    Mypy does not impose this restriction and allows an override with a plain attribute.
    `[Pyright playground] <https://pyright-play.net/?strict=true&code=MYGwhgzhAEBiD28BcBYAUNT0D6A7ArgLYBGApgE5LQCWuALuultACakBmO2t1d22ACgikQ7ADTQCJClVp0AlNAC0APmgA5eLlKoMzLMNEA6PETLloAXklmKjPZgACAB3LxnFOgE8mWNpylzIRF2RVUael19LHJSOnxyXGhDdhNAuzR7UEgYACEwcgEEeHkorHTKCIY0IA>`_
    `[mypy playground] <https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=6f860a865c5d13cce07d6cbb08b9fb85>`_

.. [#runtime]
    This PEP focuses solely on the type-checking behavior. Nevertheless, it should
    be desirable the name is read-only at runtime.

.. [#invalid_typevar]
    The implied type variable is not valid in this context; it has been used for

>     the ease of demonstration. See `ClassVar <https://typing.readthedocs.io/en/latest/spec/class-compat.html#classvar>`_.

Reviewer 3: What are we seeing `ClassVar` for here?

Author: There's this section in the linked spec:
> Note that a ClassVar parameter cannot include any type variables, regardless of the level of nesting: ClassVar[T] and ClassVar[list[set[T]]] are both invalid if T is a type variable.

Reviewer 3: Ah, I see. To be honest, I think this whole footnote can be deleted. There is nothing in the section that it is footnoting that defines `T` as specifically a typevar, requiring this disclaimer; T is just a stand-in name for "some arbitrary type name". This is very typical in typing documentation/spec and doesn't require further comment, IMO.
